### DISCUSSION TITLE
CI upstream workspace

### DISCUSSION THREAD

**Discussion by johnwason** [tesseract-robotics/tesseract_python#6] [General]
@Levi-Armstrong how does the CI handle the upstream workspace packages? Does it rebuild them, or it is pulling them from a cache? I didn't see where the cache is being pulled from, or where the upstream is cloned.

**Comment by Levi-Armstrong** [comment]
I am using the ccache action, though I am not sure if it just caches the target workspace or if it caches all three workspaces (target, upstream and downstream). Industrial CI does the cloning of each and they provide variables for each that you have access to. Now this relative to linux ci  builds the windows build does not use industrial ci.

**Reply by johnwason** [reply]
I am seeing a "upstream_ws" directory in the CI logs. Where does that come from?

**Reply by Levi-Armstrong** [reply]
It comes from this [line](https://github.com/ros-industrial-consortium/tesseract_python/blob/d195e04098ad7c2eca9e445bf015c5c6947f8eb1/.github/workflows/focal_build.yml#L21) in the CI.

**Reply by johnwason** [reply]
Ok, so I guess ccache is also catching that workspace.

**Reply by Levi-Armstrong** [reply]
I believe that is true. In the CI script you set a variable telling it where the ccache directory is so it must use this same directory for all workspaces Then the github actions saves everything in this directory. There may be a way to specify a different ccache directory for each workspace, or you could add steps to remove directories for the ccache directory before the action saves.
